File: VB/RangeDate/MyGridControl/MyDateFilterPopup.vb
' Developer Express Code Central Example:
' How to implement excel style tree filtering in a filter popup.
' 
' This example demonstrates how to filter dates using a tree and select date
' ranges.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=T143068

Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports System.Linq
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Helpers
Imports System.Collections.Generic
Imports DevExpress.XtraEditors
Imports DevExpress.Data.Filtering
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes
Imports System.Globalization
Imports DevExpress.Data.Filtering.Helpers


Namespace DateRange

	Public Class MyDateFilterPopup
		Inherits DateFilterPopup

		Private item As RepositoryItemPopupContainerEdit
		Private customCheck As CheckEdit

		Public Sub New(ByVal view As ColumnView, ByVal column As GridColumn, ByVal ownerControl As Control, ByVal creator As Object)
			MyBase.New(view, column, ownerControl, creator)

		End Sub
		Private Property treelistLocation() As Point
		Private Property treelist() As TreeList
		Private Property DateFilterControl() As PopupOutlookDateFilterControl
		Private Sub CalcControlsLocation()
			For Each ctrl As Control In DateFilterControl.Controls
				Dim ce As CheckEdit = (TryCast(ctrl, CheckEdit))
				If ce IsNot Nothing AndAlso ce.Text = View.customName Then
					treelistLocation = New Point(ctrl.Location.X, ctrl.Location.Y + 30)
				End If
			Next ctrl
		End Sub
		#Region "Creators"


		Protected Overrides Function CreateRepositoryItem() As RepositoryItemPopupBase

			item = TryCast(MyBase.CreateRepositoryItem(), RepositoryItemPopupContainerEdit)
			DateFilterControl = item.PopupControl.Controls.OfType(Of PopupOutlookDateFilterControl)().First()
			customCheck = GetCheckEdit()
			customCheck.Visible = False
			If View.treeListSource IsNot Nothing AndAlso DateFilterControl.Controls.Count > 0 Then
				CreateTreeList()
				DateFilterControl.Controls.Add(treelist)
				For Each ctrl As Control In DateFilterControl.Controls
					Dim ce As CheckEdit = (TryCast(ctrl, CheckEdit))
					If ce IsNot Nothing Then
						If ce.Text <> View.customName Then
							AddHandler ce.CheckedChanged, AddressOf OriginalDateFilterPopup_CheckedChanged
						End If
					Else
						Dim dateControlEx As DateControlEx = (TryCast(ctrl, DateControlEx))
						If dateControlEx IsNot Nothing Then
							AddHandler dateControlEx.Click, AddressOf dateControlEx_Click
						End If
					End If
				Next ctrl
				item.PopupFormMinSize = New Size(440, 280 + treelist.Height)
			End If
			Return item
		End Function
		Private Sub dateControlEx_Click(ByVal sender As Object, ByVal e As EventArgs)
			CreateDataSourceTreeList()
		End Sub
		Public Function GetCheckEdit() As CheckEdit
			For Each ctrl As Control In DateFilterControl.Controls
				If ctrl.Text = View.customName Then
					Return TryCast(ctrl, CheckEdit)
				End If
			Next ctrl
			Return Nothing
		End Function
		Private Sub CreateDataSourceTreeList()
			treelist.BeginUnboundLoad()
			treelist.ClearNodes()
			Dim parentForRootNodes As TreeListNode = Nothing
			Dim rootNode As TreeListNode = Nothing, childRootNode As TreeListNode = Nothing, childChildRootNode As TreeListNode = Nothing
			Dim filterRowArray = (TryCast(View, MyGridView)).GetFilteredValues(Me.Column, False, Nothing)
			Dim distinctYearsArray = Me.View.treeListSource.OfType(Of Date)().Select(Function(d) d.Year).Distinct().ToArray()

			For Each currentYear As Integer In distinctYearsArray
				rootNode = treelist.AppendNode(New Object() { currentYear }, parentForRootNodes)
				Dim distinctMonthArray = Me.View.treeListSource.OfType(Of Date)().Where(Function(dt) dt.Year = currentYear).Select(Function(dt) dt.ToString("MMMM")).Distinct().ToArray()
				For Each currentMonth As String In distinctMonthArray
					childRootNode = treelist.AppendNode(New Object() { currentMonth }, rootNode)
					Dim distinctDayArray = Me.View.treeListSource.OfType(Of Date)().Where(Function(dt) dt.Year = currentYear AndAlso dt.ToString("MMMM") = currentMonth).Select(Function(dt) dt.Day).Distinct().ToArray()
					For Each currentDay As Integer In distinctDayArray
						childChildRootNode = treelist.AppendNode(New Object() { currentDay }, childRootNode)
						If filterRowArray IsNot Nothing Then
							Dim currentFilter = filterRowArray.OfType(Of Date)().Where(Function(dt) (dt.Year = currentYear) AndAlso (dt.ToString("MMMM") = currentMonth) AndAlso (dt.Day = currentDay)).ToArray()
							If currentFilter.ToList().Count > 0 Then
								childChildRootNode.Checked = True
							End If
						End If
					Next currentDay
				Next currentMonth
				SetCheckNode(rootNode)
			Next currentYear
			treelist.EndUnboundLoad()
		End Sub

		Private Sub SetCheckNode(ByVal node As TreeListNode)
			Dim childCheck As Boolean
			Dim check As Boolean = True
			treelist.NodesIterator.DoLocalOperation(Sub(childNode)
				childCheck = True
				If childNode.Level = 1 Then
					treelist.NodesIterator.DoLocalOperation(Sub(childChildNode)
						If Not childChildNode.Checked Then
							childCheck = False
							check = False
						End If
					End Sub, childNode.Nodes)
					childNode.Checked = childCheck
				End If
			End Sub, node.Nodes)
			node.Checked = check
		End Sub
		Private Sub CreateTreeList()
			CalcControlsLocation()
			treelist = New TreeList()

			treelist.Location = treelistLocation

			treelist.BeginUpdate()
			treelist.Columns.Add()
			treelist.Columns(0).Caption = "Date"
			treelist.Columns(0).VisibleIndex = 0
			treelist.OptionsView.ShowCheckBoxes = True

			AddHandler treelist.AfterCheckNode, AddressOf treelist_AfterCheckNode
			treelist.EndUpdate()
			CreateDataSourceTreeList()
		End Sub
		Private Sub CreateActiveFilterCriteria(ByVal e As NodeEventArgs)
			Dim listCriteriaOperator As New List(Of CriteriaOperator)()
			treelist.NodesIterator.DoLocalOperation(Sub(node) node.CheckState = e.Node.CheckState, e.Node.Nodes)
			treelist.NodesIterator.DoOperation(Sub(node)
				If node.Level = 0 Then
					SetCheckNode(node)
					If node.Checked Then
						listCriteriaOperator.Add(GetFilterCriteriaByControlState(node))
					Else
						treelist.NodesIterator.DoLocalOperation(Sub(childNode) AddCriteria(childNode, listCriteriaOperator), node.Nodes)
					End If
				End If
			End Sub)
			Me.View.ActiveFilterCriteria = GroupOperator.Or(listCriteriaOperator)
		End Sub
		Private Sub AddCriteria(ByVal childNode As TreeListNode, ByVal listCriteriaOperator As List(Of CriteriaOperator))
			If childNode.Level = 1 Then
				If childNode.Checked Then
					listCriteriaOperator.Add(GetFilterCriteriaByControlState(childNode))
				Else
					treelist.NodesIterator.DoLocalOperation(Sub(childChildNode)
						If childChildNode.Checked Then
							listCriteriaOperator.Add(GetFilterCriteriaByControlState(childChildNode))
						End If
					End Sub, childNode.Nodes)
				End If
			End If
		End Sub
		Private Sub treelist_AfterCheckNode(ByVal sender As Object, ByVal e As NodeEventArgs)
			For Each ctrl As Control In DateFilterControl.Controls
				If ctrl.Text = View.customName Then
					Dim checkEdit As CheckEdit = CType(ctrl, CheckEdit)
					checkEdit.CheckState = CheckState.Checked
				End If
			Next ctrl
			CreateActiveFilterCriteria(e)
		End Sub

		#End Region

		Private Function GetFilterCriteriaByControlState(ByVal node As TreeListNode) As CriteriaOperator
			Return GetBetweenOperatorByName(node)
		End Function

		Protected Shadows ReadOnly Property View() As MyGridView
			Get
				Return TryCast(MyBase.View, MyGridView)
			End Get
		End Property

		Private Sub OriginalDateFilterPopup_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
			Dim checkEdit As CheckEdit = DirectCast(sender, CheckEdit)
			If checkEdit.Checked AndAlso checkEdit.Text <> "Filter by a specific date:" Then
				CreateDataSourceTreeList()
			End If
		End Sub
		Protected Overridable Function GetBetweenOperatorByName(ByVal node As TreeListNode) As BetweenOperator
			If node.Level = 0 Then
				Dim firstDay As New Date(CInt(Math.Truncate(node.GetValue(treelist.Columns(0)))), 01, 01)
				Dim endDay As New Date(CInt(Math.Truncate(node.GetValue(treelist.Columns(0)))), 12, 31)
				Return New BetweenOperator(Me.Column.FieldName, firstDay, endDay)
			ElseIf node.Level = 1 Then
				Dim firstDay As New Date(CInt(Math.Truncate(node.ParentNode.GetValue(treelist.Columns(0)))), GetMonth(node), 01)
				Dim endDay As New Date(CInt(Math.Truncate(node.ParentNode.GetValue(treelist.Columns(0)))), GetMonth(node), Date.DaysInMonth(CInt(Math.Truncate(node.ParentNode.GetValue(treelist.Columns(0)))), GetMonth(node)))
				Return New BetweenOperator(Me.Column.FieldName, firstDay, endDay)
			Else
				Dim firstDay As New Date(CInt(Math.Truncate(node.ParentNode.ParentNode.GetValue(treelist.Columns(0)))), GetMonth(node.ParentNode), CInt(Math.Truncate(node.GetValue(treelist.Columns(0)))))
				Dim endDay As New Date(CInt(Math.Truncate(node.ParentNode.ParentNode.GetValue(treelist.Columns(0)))), GetMonth(node.ParentNode), CInt(Math.Truncate(node.GetValue(treelist.Columns(0)))))
				Return New BetweenOperator(Me.Column.FieldName, firstDay, endDay)
			End If
		End Function
		Private Function GetMonth(ByVal node As TreeListNode) As Integer
			Dim dt As Date = Date.ParseExact(CStr(node.GetValue(treelist.Columns(0))), "MMMM", CultureInfo.InvariantCulture)
			Return dt.Month
		End Function
	End Class
End Namespace

File: CS/RangeDate/MyGridControl/MyDateFilterPopup.cs
// Developer Express Code Central Example:
// How to implement excel style tree filtering in a filter popup.
// 
// This example demonstrates how to filter dates using a tree and select date
// ranges.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=T143068

using System;
using System.Drawing;
using System.Windows.Forms;
using System.Linq;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Helpers;
using System.Collections.Generic;
using DevExpress.XtraEditors;
using DevExpress.Data.Filtering;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;
using System.Globalization;
using DevExpress.Data.Filtering.Helpers;


namespace DateRange
{

    public class MyDateFilterPopup : DateFilterPopup
    {
        RepositoryItemPopupContainerEdit item;
        CheckEdit customCheck;

        public MyDateFilterPopup(ColumnView view, GridColumn column, Control ownerControl, object creator)
            : base(view, column, ownerControl, creator)
        {

        }
        Point treelistLocation { get; set; }
        TreeList treelist { get; set; }
        PopupOutlookDateFilterControl DateFilterControl
        {
            get;
            set;
        }
        private void CalcControlsLocation()
        {
            foreach (Control ctrl in DateFilterControl.Controls)
            {
                CheckEdit ce = (ctrl as CheckEdit);
                if (ce != null && ce.Text == View.customName)
                    treelistLocation = new Point(ctrl.Location.X, ctrl.Location.Y + 30);
            }
        }
        #region Creators
        
       
        protected override RepositoryItemPopupBase CreateRepositoryItem()
        {

            item = base.CreateRepositoryItem() as RepositoryItemPopupContainerEdit;
            DateFilterControl = item.PopupControl.Controls.OfType<PopupOutlookDateFilterControl>().First();
            customCheck = GetCheckEdit();
            customCheck.Visible = false;
            if (View.treeListSource != null && DateFilterControl.Controls.Count > 0)
            {
                CreateTreeList(); 
                DateFilterControl.Controls.Add(treelist);
                foreach (Control ctrl in DateFilterControl.Controls)
                {
                    CheckEdit ce = (ctrl as CheckEdit);
                    if (ce != null)
                    {
                        if (ce.Text != View.customName)
                            ce.CheckedChanged += OriginalDateFilterPopup_CheckedChanged;
                    }
                    else
                    {
                        DateControlEx dateControlEx = (ctrl as DateControlEx);
                        if (dateControlEx != null)
                        {
                            dateControlEx.Click += dateControlEx_Click;
                        }
                    }
                }
                item.PopupFormMinSize = new Size(440, 280 + treelist.Height);
            }
            return item;
        }
        private void dateControlEx_Click(object sender, EventArgs e)
        {
            CreateDataSourceTreeList();
        }
        public CheckEdit GetCheckEdit()
        {
            foreach (Control ctrl in DateFilterControl.Controls)
            {
                if (ctrl.Text == View.customName)
                    return ctrl as CheckEdit;
            }
            return null;
        }
        private void CreateDataSourceTreeList()
        {
            treelist.BeginUnboundLoad();
            treelist.ClearNodes();
            TreeListNode parentForRootNodes = null;
            TreeListNode rootNode = null,
                childRootNode = null,
                childChildRootNode = null;
            var filterRowArray = (View as MyGridView).GetFilteredValues(this.Column, false, null);
            var distinctYearsArray = this.View.treeListSource.OfType<DateTime>().Select((d) => d.Year).Distinct().ToArray();

            foreach (int currentYear in distinctYearsArray)
            {
                rootNode = treelist.AppendNode(new object[] { currentYear }, parentForRootNodes);
                var distinctMonthArray = this.View.treeListSource.OfType<DateTime>().Where((dt) => dt.Year == currentYear).Select((dt) => dt.ToString("MMMM")).Distinct().ToArray();
                foreach (string currentMonth in distinctMonthArray)
                {
                    childRootNode = treelist.AppendNode(new object[] { currentMonth }, rootNode);
                    var distinctDayArray = this.View.treeListSource.OfType<DateTime>().Where((dt) => dt.Year == currentYear && dt.ToString("MMMM") == currentMonth).Select((dt) => dt.Day).Distinct().ToArray();
                    foreach (int currentDay in distinctDayArray)
                    {
                        childChildRootNode = treelist.AppendNode(new object[] { currentDay }, childRootNode);
                        if (filterRowArray != null)
                        {
                            var currentFilter = filterRowArray.OfType<DateTime>().Where((dt) => (dt.Year == currentYear) && (dt.ToString("MMMM") == currentMonth) && (dt.Day == currentDay)).ToArray();
                            if (currentFilter.ToList().Count > 0)
                                childChildRootNode.Checked = true;
                        }
                    }
                }
                SetCheckNode(rootNode);
            }
            treelist.EndUnboundLoad();
        }
        
        private void SetCheckNode(TreeListNode node)
        {
            bool childCheck;
            bool check = true;
            treelist.NodesIterator.DoLocalOperation((childNode) => 
            {
                childCheck = true;
                if (childNode.Level == 1)
                {
                    treelist.NodesIterator.DoLocalOperation((childChildNode) =>
                       {
                           if (!childChildNode.Checked)
                           {
                               childCheck = false;
                               check = false;
                           }
                       }, childNode.Nodes);
                    childNode.Checked = childCheck;
                }
            }, node.Nodes);
            node.Checked = check;
        }
        private void CreateTreeList()
        {
            CalcControlsLocation();
            treelist = new TreeList();

            treelist.Location = treelistLocation;

            treelist.BeginUpdate();
            treelist.Columns.Add();
            treelist.Columns[0].Caption = "Date";
            treelist.Columns[0].VisibleIndex = 0;
            treelist.OptionsView.ShowCheckBoxes = true;

            treelist.AfterCheckNode += treelist_AfterCheckNode;
            treelist.EndUpdate();
            CreateDataSourceTreeList();
        }
        private void CreateActiveFilterCriteria(NodeEventArgs e)
        {
            List<CriteriaOperator> listCriteriaOperator = new List<CriteriaOperator>();
            treelist.NodesIterator.DoLocalOperation((node) => { 
                node.CheckState = e.Node.CheckState; 
            }
            , e.Node.Nodes);
            treelist.NodesIterator.DoOperation((node) =>
            {
                if (node.Level == 0)
                { 
                   SetCheckNode(node);
                    if (node.Checked)
                        listCriteriaOperator.Add(GetFilterCriteriaByControlState(node));
                    else
                    {
                        treelist.NodesIterator.DoLocalOperation((childNode) =>
                        {
                            AddCriteria(childNode, listCriteriaOperator);
                        }, node.Nodes);
                    }
                }
            });       
            this.View.ActiveFilterCriteria = GroupOperator.Or(listCriteriaOperator);
        }
        private void AddCriteria(TreeListNode childNode, List<CriteriaOperator> listCriteriaOperator)
        {
            if (childNode.Level == 1)
            {
                if (childNode.Checked)
                    listCriteriaOperator.Add(GetFilterCriteriaByControlState(childNode));
                else
                    treelist.NodesIterator.DoLocalOperation((childChildNode) =>
                    {
                        if (childChildNode.Checked)
                            listCriteriaOperator.Add(GetFilterCriteriaByControlState(childChildNode));
                    }, childNode.Nodes);
            }
        }
        void treelist_AfterCheckNode(object sender, NodeEventArgs e)
        {
            foreach (Control ctrl in DateFilterControl.Controls)
            {
                if (ctrl.Text == View.customName)
                {
                    CheckEdit checkEdit = (CheckEdit)ctrl;
                    checkEdit.CheckState = CheckState.Checked;
                }
            }
            CreateActiveFilterCriteria(e);
        }

        #endregion

        private CriteriaOperator GetFilterCriteriaByControlState(TreeListNode node)
        {
            return GetBetweenOperatorByName(node);
        }

        protected new MyGridView View
        {
            get { return base.View as MyGridView; }
        }

        void OriginalDateFilterPopup_CheckedChanged(object sender, EventArgs e)
        {
            CheckEdit checkEdit = (CheckEdit)sender;
            if (checkEdit.Checked && checkEdit.Text != "Filter by a specific date:")
                CreateDataSourceTreeList();
        }
        protected virtual BetweenOperator GetBetweenOperatorByName(TreeListNode node)
        {
            if (node.Level == 0)
            {
                DateTime firstDay = new DateTime((int)node.GetValue(treelist.Columns[0]), 01, 01);
                DateTime endDay = new DateTime((int)node.GetValue(treelist.Columns[0]), 12, 31);
                return new BetweenOperator(this.Column.FieldName, firstDay, endDay);
            }
            else if (node.Level == 1)
            {
                DateTime firstDay = new DateTime((int)node.ParentNode.GetValue(treelist.Columns[0]), GetMonth(node), 01);
                DateTime endDay = new DateTime((int)node.ParentNode.GetValue(treelist.Columns[0]), GetMonth(node), DateTime.DaysInMonth((int)node.ParentNode.GetValue(treelist.Columns[0]), GetMonth(node)));
                return new BetweenOperator(this.Column.FieldName, firstDay, endDay);
            }
            else
            {
                DateTime firstDay = new DateTime((int)node.ParentNode.ParentNode.GetValue(treelist.Columns[0]), GetMonth(node.ParentNode), (int)node.GetValue(treelist.Columns[0]));
                DateTime endDay = new DateTime((int)node.ParentNode.ParentNode.GetValue(treelist.Columns[0]), GetMonth(node.ParentNode), (int)node.GetValue(treelist.Columns[0]));
                return new BetweenOperator(this.Column.FieldName, firstDay, endDay);
            }
        }
        private int GetMonth(TreeListNode node)
        {
            DateTime dt = DateTime.ParseExact((string)node.GetValue(treelist.Columns[0]), "MMMM", CultureInfo.InvariantCulture);
            return dt.Month;
        }
    }
}